AsyncUsageAnalyzers/AsyncUsageAnalyzers.CodeFixes/Usage/DontUseThreadSleepCodeUniversalCodeFixProvider.cs

```diff
@@ -0,0 +1,160 @@
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Text;
+
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(DontUseThreadSleepCodeUniversalCodeFixProvider))]
+    [Shared]
+    internal class DontUseThreadSleepCodeUniversalCodeFixProvider : CodeFixProvider
+    {
+        private static readonly ImmutableArray<string> FixableDiagnostics =
+                ImmutableArray.Create(DontUseThreadSleepAnalyzer.DiagnosticId, DontUseThreadSleepInAsyncCodeAnalyzer.DiagnosticId);
+
+        public override ImmutableArray<string> FixableDiagnosticIds => FixableDiagnostics;
+
+        /// <inheritdoc/>
+        public override FixAllProvider GetFixAllProvider()
+        {
+            return CustomFixAllProviders.BatchFixer;
+        }
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics)
+            {
+                if (diagnostic.Id == DontUseThreadSleepInAsyncCodeAnalyzer.DiagnosticId)
+                {
+                    RegisterCodeFixForDiagnostic(context, diagnostic);
+                }
+                else if (diagnostic.Id == DontUseThreadSleepAnalyzer.DiagnosticId)
+                {
+                    var document = context.Document;
+
+                    var root = await document.GetSyntaxRootAsync().ConfigureAwait(false);
+                    var invocationExpression = root.FindNode(TextSpan.FromBounds(diagnostic.Location.SourceSpan.Start, diagnostic.Location.SourceSpan.End), getInnermostNodeForTie: true) as InvocationExpressionSyntax;
+
+                    if (invocationExpression == null)
+                    {
+                        return;
+                    }
+
+                    if (invocationExpression.IsInsideAsyncCode())
+                    {
+                        RegisterCodeFixForDiagnostic(context, diagnostic);
+                    }
+                }
+            }
+        }
+
+        private static void RegisterCodeFixForDiagnostic(CodeFixContext context, Diagnostic diagnostic)
+        {
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    "Use `await Task.Delay(...)` or `await Task.Yield()`",
+                    cancellationToken => GetTransformedDocumentAsync(context.Document, diagnostic, cancellationToken)),
+                diagnostic);
+        }
+
+        private static async Task<Document> GetTransformedDocumentAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var firstNodeWithCorrectSpan = root
+                .FindNode(diagnostic.Location.SourceSpan);
+            InvocationExpressionSyntax expression = firstNodeWithCorrectSpan
+                .DescendantNodesAndSelf()
```

Comment:
I don't know Roslyn very well either, but I think that I can justify that `expression ` and `firstNodeWithCorrectSpan` are not null. 

1. `DontUseThreadSleepAnalyzerBase` is registered on `SyntaxKind.InvocationExpression` and reports analysis on `InvocationExpression.Syntax`'s location
2.  I use diangnostic's location to assign value to`firstNodeWithCorrectSpan`. `firstNodeWithCorrectSpan` is assigned using `SyntaxNode.FindNode` method which (by default) returns outermost node encompassing the given span (source: http://www.coderesx.com/roslyn/html/6497A5DA.htm).
3. There must be a node of type `InvocationExpressionSyntax` among `firstNodeWithCorrectSpan`'s descendant nodes because diagnostic was reported on using `InvocationExpressionSyntax`'s location.